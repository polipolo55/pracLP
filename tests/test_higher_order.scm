(define (power base exponent)
  ;; Calcula base^exponent recursivament
  (if (= exponent 0)
      1
      (* base (power base (- exponent 1)))))

(define (apply-twice f val)
  ;; Exemple de funció d'ordre superior
  (f (f val)))

(define (is-even? n)
  (if (= n 0)
      #t
      (if (= n 1)
          #f
          (is-even? (- n 2)))))

(define (factorial n)
  (if (<= n 1)
      1
      (* n (factorial (- n 1)))))

(define (fibonacci n)
  (if (< n 2)
      n
      (+ (fibonacci (- n 1)) (fibonacci (- n 2)))))


(define (main)
  (display "Prova de funció d'ordre superior 'apply-twice' amb 'factorial':")
  (newline)
  (display "apply-twice factorial 3: ")
  (display (apply-twice factorial 3))
  (newline)


  (display "Prova de passatge de funcions com a paràmetres:")
  (newline)
  (display "Cridant 'apply-twice' amb 'is-even?' i 4: ")
  (display (apply-twice is-even? 4))
  (newline)

  (display "Totes les proves de funcions d'ordre superior han finalitzat!")
  (newline)
)