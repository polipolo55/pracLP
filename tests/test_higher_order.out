Prova de funció d'ordre superior 'apply-twice' amb 'factorial':

apply-twice factorial 3: 
720

Prova de passatge de funcions com a paràmetres:

Cridant 'apply-twice' amb 'is-even?' i 4: 
False

Totes les proves de funcions d'ordre superior han finalitzat!

